### PR TITLE
Create a `MockApplication` test helper type

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

There are a few tests in `linera-execution` that require an application instance. This was the reason `TestApplication` was developed. As more tests were added, `TestApplication` started gaining more "features", and it was hard to understand which tests were using which features from the `TestApplication`. More importantly, it was getting complicated to add more tests because they usually required adding new features to the `TestApplication` without breaking any of the existing features.

## Proposal

Replace the `TestApplication` with a `MockApplication` which can be customized for each individual unit test. Refactor the unit tests to use it.

## Test Plan

The tests are still the same, just refactored.

## Release Plan

Internal refactoring of tests, nothing needed.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-execution/tests/test_execution.rs [hunk 1]
```diff
@@ -446,32 +278,128 @@ async fn test_simple_user_operation_with_leaking_session(
     Ok(())
 }
 
+/// Tests if a session is called correctly during execution.
+#[tokio::test]
+async fn test_simple_session() -> anyhow::Result<()> {
```
Comment (jvff) [inline linera-execution/tests/test_execution.rs:283]: This is a new test for the "happy path" of a simple session.

File: linera-execution/tests/test_execution.rs [hunk 2]
```diff
@@ -64,289 +68,92 @@ async fn test_missing_bytecode_for_user_application() -> anyhow::Result<()> {
...
-        }))
-    }
-}
-
-/// Key to store the application ID to call.
-static CALLEE_ID_KEY: [u8; 1] = [0];
-/// Key to store some dummy data in the application state.
-static DUMMY_STATE_KEY: [u8; 1] = [1];
+#[tokio::test]
+async fn test_simple_user_operation() -> anyhow::Result<()> {
```
Comment (jvff) [inline linera-execution/tests/test_execution.rs:72]: This test now seems to be testing a few different things, including cross-application calls, authentication forwarding, session calls, and runtime storage. I'm thinking about adding a `TODO` here to split this test into smaller and more specialized tests. But I'd like to know your opinion about this and the overall refactor before opening the issue :sweat_smile:

File: linera-execution/tests/test_execution.rs [hunk 3]
```diff
@@ -446,32 +278,128 @@ async fn test_simple_user_operation_with_leaking_session(
...
+            Operation::User {
+                application_id: caller_id,
+                bytes: vec![],
+            },
+            &policy,
+            &mut tracker,
+        )
+        .await;
+
+    assert!(result.is_ok());
```
Comment (jvff) [inline linera-execution/tests/test_execution.rs:351]: This is not great. I'll try to fix it.

Comment (jvff) [reply linera-execution/tests/test_execution.rs]: Done.

**Review by ma2bd — APPROVED** [review]
Very nice!

File: linera-execution/tests/test_execution.rs (hunk 2 above)
Comment (ma2bd) [reply linera-execution/tests/test_execution.rs:72]: Yes, I would suggest adding a TODO pointing a GH issue labeled as a "good first task".

**Pull request merged**